File: app/src/main/java/com/wiseassblog/samsaradayplanner/ui/managetaskview/TaskViewLogic.java
package com.wiseassblog.samsaradayplanner.ui.managetaskview;

import com.wiseassblog.samsaradayplanner.common.BaseViewLogic;
import com.wiseassblog.samsaradayplanner.common.Continuation;
import com.wiseassblog.samsaradayplanner.domain.ITaskStorage;
import com.wiseassblog.samsaradayplanner.domain.Task;
import com.wiseassblog.samsaradayplanner.domain.constants.COLOR;
import com.wiseassblog.samsaradayplanner.domain.constants.ICON;

public class TaskViewLogic extends BaseViewLogic<TaskViewEvent> {

    private ITaskViewContract.View view;
    private ITaskViewContract.ViewModel vm;
    private ITaskStorage storage;

    /**
     * Why skip Use Cases/Interactors/Transaction Scripts and just call the Repository directly?
     * Because this feature is so simple that adding in the domain layer (Use Case...etc.) is
     * not worth the effort.
     *
     * @param view
     * @param vm
     * @param storage
     */
    public TaskViewLogic(ITaskViewContract.View view, ITaskViewContract.ViewModel vm, ITaskStorage storage) {
        this.view = view;
        this.vm = vm;
        this.storage = storage;
    }

    @Override
    public void onViewEvent(TaskViewEvent event) {
        //code like this make me miss Kotlin syntax...
        switch (event.getEvent()) {
            case ON_START:
                onStart();
                break;
            case ON_COLOR_BUTTON_CLICK:
                view.showColorPickerSheet();
                break;
            case ON_COLOR_SELECTED:
                onColorSelected((COLOR) event.getValue());
                break;
            case ON_DONE_CLICK:
                updateStorage();
                break;
            case ON_ICON_SELECTED:
                onIconSelected(
                        (ICON) event.getValue()
                );
                break;
        }
    }


    private void onStart() {
        storage.getTask(vm.getTask().getTaskId(),
                new Continuation<Task>() {
                    @Override
                    public void onSuccess(Task result) {
                        vm.setTask(result);
                        view.setButtonColor(result.getTaskColor());
                        view.setName(result.getTaskName());
                        view.setIconSelection(result.getTaskIcon());
                        view.setSelectedSpinnerItem(result.getTaskId());
                    }

                    @Override
                    public void onException(Exception e) {
                        view.showMessage(e.getMessage());
                        view.goToTaskListActivity();
                    }
                }
        );
    }

    private void onIconSelected(ICON icon) {
        Task oldTask = vm.getTask();
        Task update = new Task(
                oldTask.getTaskId(),
                oldTask.getTaskName(),
                icon,
                oldTask.getTaskColor()
        );

        vm.setTask(update);
        view.setIconSelection(update.getTaskIcon());
    }

    private void updateStorage() {
        Task oldTask = vm.getTask();
        Task update = new Task(
                oldTask.getTaskId(),
                view.getName(),
                oldTask.getTaskIcon(),
                oldTask.getTaskColor()
        );

        storage.updateTask(update,
                new Continuation<Void>() {
                    @Override
                    public void onSuccess(Void result) {
                        view.goToTaskListActivity();
                    }

                    @Override
                    public void onException(Exception e) {
                        view.showMessage(e.getMessage());
                        view.goToTaskListActivity();
                    }
                }
        );
    }

    private void onColorSelected(COLOR color) {
        Task oldTask = vm.getTask();
        Task newTask = new Task(
                oldTask.getTaskId(),
                oldTask.getTaskName(),
                oldTask.getTaskIcon(),
                color
        );
        vm.setTask(
                newTask
        );

        view.setButtonColor(newTask.getTaskColor());
        view.hideColorPickerSheet();
    }


}

File: app/src/main/java/com/wiseassblog/samsaradayplanner/ui/managetaskview/TaskViewLogic.kt
package com.wiseassblog.samsaradayplanner.ui.managetaskview

import com.wiseassblog.samsaradayplanner.common.BaseViewLogic
import com.wiseassblog.samsaradayplanner.domain.ITaskStorage
import com.wiseassblog.samsaradayplanner.domain.Task
import com.wiseassblog.samsaradayplanner.domain.constants.COLOR
import com.wiseassblog.samsaradayplanner.domain.constants.ICON
import com.wiseassblog.samsaradayplanner.domain.constants.Messages.GENERIC_ERROR_MESSAGE

class TaskViewLogic
/**
 * Why skip Use Cases/Interactors/Transaction Scripts and just call the Repository directly?
 * Because this feature is so simple that adding in the domain layer (Use Case...etc.) is
 * not worth the effort.
 *
 * @param view
 * @param vm
 * @param storage
 */(
    private val view: ITaskViewContract.View,
    private val vm: ITaskViewContract.ViewModel,
    private val storage: ITaskStorage
) : BaseViewLogic<TaskViewEvent>() {
    override fun onViewEvent(event: TaskViewEvent) {
        when (event) {
            is TaskViewEvent.OnStart -> onStart()
            is TaskViewEvent.OnColorButtonClick -> view.showColorPickerSheet()
            is TaskViewEvent.OnColorSelected -> onColorSelected(event.color)
            is TaskViewEvent.OnDoneClick -> updateStorage()
            is TaskViewEvent.OnIconSelected -> onIconSelected(event.icon)
        }
    }

    private fun onStart() {
        storage.getTask(
            vm.task.taskId,
            { task ->
                vm.task = task
                view.setButtonColor(task.taskColor)
                view.name = task.taskName
                view.setIconSelection(task.taskIcon)
                view.setSelectedSpinnerItem(task.taskId)
            },
            {
                view.showMessage(GENERIC_ERROR_MESSAGE)
                view.goToTaskListActivity()
            }
        )
    }

    private fun onIconSelected(icon: ICON) {
        val oldTask = vm.task
        val update = Task(
            oldTask.taskId,
            oldTask.taskName,
            icon,
            oldTask.taskColor
        )
        vm.task = update
        view.setIconSelection(update.taskIcon)
    }

    private fun updateStorage() {
        val oldTask = vm.task
        val update = Task(
            oldTask.taskId,
            view.name,
            oldTask.taskIcon,
            oldTask.taskColor
        )
        storage.updateTask(
            update,
            {
                view.goToTaskListActivity()
            },
            {
                view.showMessage(GENERIC_ERROR_MESSAGE)
                view.goToTaskListActivity()
            }

        )
    }

    private fun onColorSelected(color: COLOR) {
        val oldTask = vm.task
        val newTask = Task(
            oldTask.taskId,
            oldTask.taskName,
            oldTask.taskIcon,
            color
        )
        vm.task = newTask
        view.setButtonColor(newTask.taskColor)
        view.hideColorPickerSheet()
    }
}